In [93]:
import pandas as pd

In [94]:
w_dummy_results = pd.read_csv('../docs_&_notes/first_ann_run_results.csv', delimiter=';')

In [95]:
w_dummy_results.head()

loss_function optimizer columns_used  batch_size  num_epochs  \
0  binary_crossentropy     nadam     elo_diff           8          50   
1  binary_crossentropy     nadam     elo_diff           8         100   
2  binary_crossentropy     nadam     elo_diff           8         200   
3  binary_crossentropy     nadam     elo_diff          20          50   
4  binary_crossentropy     nadam     elo_diff          20         100   

   confusion_matrix  accuracy  
0  [30, 18][14, 38]        68  
1  [26, 22][12, 40]        66  
2  [30, 18][14, 38]        68  
3  [26, 22][12, 40]        66  
4  [26, 22][12, 40]        66

In [96]:
w_dummy_results.describe()

batch_size   num_epochs     accuracy
count  2627.000000  2627.000000  2627.000000
mean     38.687476   119.375714    68.380282
std      31.502913    62.670259     2.761330
min       8.000000    50.000000    57.000000
25%       8.000000    50.000000    66.000000
50%      20.000000   100.000000    68.000000
75%      44.000000   200.000000    70.000000
max      92.000000   200.000000    77.000000

In [97]:
w_dummy_results.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2627 entries, 0 to 2626
Data columns (total 7 columns):
loss_function       2627 non-null object
optimizer           2627 non-null object
columns_used        2627 non-null object
batch_size          2627 non-null int64
num_epochs          2627 non-null int64
confusion_matrix    2627 non-null object
accuracy            2627 non-null int64
dtypes: int64(3), object(4)
memory usage: 143.7+ KB


### Let's see which columns give the best average accuracy 

In [98]:
grouped_by_cols = w_dummy_results.groupby(by='columns_used').mean()
grouped_by_cols.sort_values('accuracy', ascending=False)

batch_size  num_epochs  \
columns_used                                                              
elo_diff, start_time, weekday                     38.804878  119.817073   
elo_diff, start_time, game_time, weekday          38.690909  119.393939   
elo_diff, color, start_time, game_time, weekday   38.690909  119.393939   
elo_diff, game_time                               38.690909  119.393939   
elo_diff                                          38.690909  119.393939   
elo_diff, color, start_time, weekday              38.690909  119.393939   
elo_diff, start_time, game_time                   38.690909  119.393939   
elo_diff, weekday                                 38.690909  119.393939   
elo_diff, game_time, weekday                      38.690909  118.484848   
elo_diff, color, start_time, game_time            38.690909  119.393939   
elo_diff, start_time                              38.690909  119.393939   
elo_diff, color, start_time                       38.690909  119.393939   
elo_diff, color, game_time                        38.690909  119.393939   
elo_diff, color                                   38.690909  119.393939   
elo_diff, color, game_time, weekday               38.509804  119.607843   
elo_diff, color, weekday                          38.690909  119.393939   

                                                  accuracy  
columns_used                                                
elo_diff, start_time, weekday                    71.451220  
elo_diff, start_time, game_time, weekday         70.775758  
elo_diff, color, start_time, game_time, weekday  69.424242  
elo_diff, game_time                              69.406061  
elo_diff                                         69.309091  
elo_diff, color, start_time, weekday             69.169697  
elo_diff, start_time, game_time                  68.903030  
elo_diff, weekday                                68.545455  
elo_diff, game_time, weekday                     68.472727  
elo_diff, color, start_time, game_time           68.375758  
elo_diff, start_time                             67.763636  
elo_diff, color, start_time                      66.993939  
elo_diff, color, game_time                       66.812121  
elo_diff, color                                  66.357576  
elo_diff, color, game_time, weekday              66.267974  
elo_diff, color, weekday                         65.921212

### Winners and Loser (+Fringers?):
#### Winners (Greater than 70%):
- elo_diff, start_time, weekday = 71.5%
- elo_diff, start_time, game_time, weekday = 70.8%

#### Losers (Less than 2/3 (66.67%)):
- elo_diff, color, weekday = 65.9%
- elo_diff, color, game_time, weekday = 66.3%
- elo_diff, color = 66.4%

#### Winners on the fringe (Above 69%, less than 70%):
- elo_diff, color, start_time, game_time, weekday = 69.4%
- elo_diff, game_time = 69.4%
- elo_diff = 69.3%
- elo_diff, color, start_time, weekday = 69.2%

#### Losers on the fringe (Above 66.67%, less than 68%):
- elo_diff, color, game_time = 66.8%
- elo_diff, color, start_time = 67.0%
- elo_diff, start_time = 67.8%

#### In-Betweeners (Between 68% and 69%):
- elo_diff, start_time, game_time = 68.9%
- elo_diff, weekday = 68.5%
- elo_diff, game_time, weekday = 68.5%
- elo_diff, color, start_time, game_time = 68.4%

### Real Test! Group columns by the column name, loss function, and optimizer

In [101]:
grouped_by_cols = w_dummy_results.groupby(by=['columns_used', 'loss_function', 'optimizer']).mean()
grouped_df = grouped_by_cols.sort_values(by='accuracy', ascending=False).reset_index()

In [113]:
grouped_df[grouped_df['accuracy']>=72]

columns_used        loss_function optimizer  \
0             elo_diff, start_time, weekday                  mae      adam   
1             elo_diff, start_time, weekday                  mae     nadam   
2  elo_diff, start_time, game_time, weekday                  mae     nadam   
3             elo_diff, start_time, weekday  binary_crossentropy   adagrad   
4  elo_diff, start_time, game_time, weekday                  mse  adadelta   
5  elo_diff, start_time, game_time, weekday                  mae  adadelta   
6  elo_diff, start_time, game_time, weekday  binary_crossentropy   adagrad   
7  elo_diff, start_time, game_time, weekday                  mae      adam   
8             elo_diff, start_time, weekday                  mae   adagrad   
9             elo_diff, start_time, weekday                  mse   adagrad   

   batch_size  num_epochs   accuracy  
0        41.0  116.666667  73.000000  
1        41.0  116.666667  72.916667  
2        41.0  116.666667  72.750000  
3        41.0  116.666667  72.666667  
4        18.0  141.666667  72.500000  
5        18.0  141.666667  72.500000  
6        41.0  116.666667  72.500000  
7        41.0  116.666667  72.333333  
8        41.0  116.666667  72.083333  
9        41.0  116.666667  72.083333

In [116]:
grouped_df[grouped_df['accuracy']<66]

columns_used        loss_function optimizer  \
214                       elo_diff, color                  mae     nadam   
215              elo_diff, color, weekday                  mse      adam   
216              elo_diff, color, weekday                  mae   adagrad   
217   elo_diff, color, game_time, weekday                  mse     nadam   
218           elo_diff, color, start_time  binary_crossentropy     nadam   
219              elo_diff, color, weekday  binary_crossentropy   adagrad   
220              elo_diff, color, weekday  binary_crossentropy   rmsprop   
221   elo_diff, color, game_time, weekday                  mae   rmsprop   
222   elo_diff, color, game_time, weekday  binary_crossentropy   adagrad   
223          elo_diff, game_time, weekday                  mse     nadam   
224  elo_diff, color, start_time, weekday                  mse   adagrad   
225              elo_diff, color, weekday                  mae   rmsprop   
226           elo_diff, color, start_time                  mae      adam   
227   elo_diff, color, game_time, weekday                  mse   adagrad   
228                       elo_diff, color  binary_crossentropy     nadam   
229   elo_diff, color, game_time, weekday                  mae   adagrad   
230              elo_diff, color, weekday  binary_crossentropy     nadam   
231            elo_diff, color, game_time                  mse     nadam   
232                       elo_diff, color                  mse     nadam   
233              elo_diff, color, weekday                  mse     nadam   
234            elo_diff, color, game_time  binary_crossentropy     nadam   
235            elo_diff, color, game_time                  mae   adagrad   
236                       elo_diff, color                  mae      adam   
237                       elo_diff, color                  mae   rmsprop   
238                       elo_diff, color                  mae   adagrad   

     batch_size  num_epochs   accuracy  
214        41.0  116.666667  65.833333  
215        41.0  116.666667  65.833333  
216        41.0  116.666667  65.833333  
217        41.0  116.666667  65.750000  
218        41.0  116.666667  65.750000  
219        41.0  116.666667  65.750000  
220        41.0  116.666667  65.666667  
221        41.0  116.666667  65.583333  
222        41.0  116.666667  65.583333  
223        41.0  116.666667  65.583333  
224        41.0  116.666667  65.500000  
225        41.0  116.666667  65.500000  
226        41.0  116.666667  65.500000  
227        41.0  116.666667  65.500000  
228        41.0  116.666667  65.333333  
229        41.0  116.666667  65.250000  
230        41.0  116.666667  65.083333  
231        41.0  116.666667  65.000000  
232        41.0  116.666667  64.916667  
233        41.0  116.666667  64.833333  
234        41.0  116.666667  64.666667  
235        41.0  116.666667  64.583333  
236        41.0  116.666667  64.500000  
237        41.0  116.666667  64.250000  
238        41.0  116.666667  63.833333

In [109]:
print('BY Column Name:\nLess than 67%:')
print(grouped_df[grouped_df['accuracy']<=67].columns_used.value_counts())
print('\n\nBY Column Name:\nGreater than 71%:')
print(grouped_df[grouped_df['accuracy']>=71].columns_used.value_counts())
print('\n\nBY Loss Function:\nLess than 67%:')
print(grouped_df[grouped_df['accuracy']<=67].loss_function.value_counts())
print('\n\nBY Loss Function:\nGreater than 71%:')
print(grouped_df[grouped_df['accuracy']>=71].loss_function.value_counts())
print('\n\nBY Optimizer:\nLess than 67%:')
print(grouped_df[grouped_df['accuracy']<=67].optimizer.value_counts())
print('\n\nBY Optimizer:\nGreater than 71%:')
print(grouped_df[grouped_df['accuracy']>=71].optimizer.value_counts())

BY Column Name:
Less than 67%:
elo_diff, color, weekday                13
elo_diff, color, game_time, weekday     10
elo_diff, color                          9
elo_diff, color, game_time               7
elo_diff, color, start_time              5
elo_diff, weekday                        3
elo_diff, game_time, weekday             2
elo_diff, color, start_time, weekday     2
elo_diff, start_time, game_time          1
elo_diff                                 1
elo_diff, start_time                     1
Name: columns_used, dtype: int64


BY Column Name:
Greater than 71%:
elo_diff, start_time, game_time, weekday    9
elo_diff, start_time, weekday               9
elo_diff, color, start_time, weekday        2
elo_diff, start_time, game_time             2
elo_diff                                    2
elo_diff, game_time                         1
Name: columns_used, dtype: int64


BY Loss Function:
Less than 67%:
mae                    19
mse                    18
binary_crossentropy    17
Name:

## Conclusions:

1. Assigned color doesn't make much difference if the result is going to be a win or a loss
2. Length of time assigned to players also has little to no impact in the outcome of the game
3. Time of the day and day of the week makes a much larger impact in determining who wins
4. Mean Absolute Error and Mean Square Error work for classification model too
5. AdaDelta is called "more refined" version of AdaGrad. But, AdaDelta only works if enough epochs are being run for a batch
6. Nadam, Adam, and RMS-Prop are great, if given the right parameters

### Now it is time to run Grid Search on the top two combinations.
1. elo_diff + start_time + weekday
2. elo_diff + start_time + game_time + weekday